File: 1466/reorder-routes-to-make-all-paths-lead-to-the-city-zero.java
class Solution { //條條大路通羅馬
    public int minReorder(int n, int[][] connections) {
        //題目描述，edge有n-1，所以其實就是最精簡的連線狀況。
        //所以解法就是，從0開始，進行 graph 旅行。然後看前後關係反過來的有幾個，就是答案
        //資料結構，可以存 connections的index
        boolean [] visited = new boolean[n];
        //ArrayList<ArrayList<Integer>> table = new ArrayList<ArrayList<Integer>>();
        ArrayList<Integer> [] table = new ArrayList[n];
        for(int i=0; i<n; i++){
            //table.add(new ArrayList<Integer>());
            table[i] = new ArrayList<Integer>();
        }
        
        for(int i=0; i<connections.length; i++){
            int a = connections[i][0], b = connections[i][1];
            //table.get(a).add(i);
            //table.get(b).add(i);
            table[a].add(i);
            table[b].add(i);
        }

        int ans = 0;
        LinkedList<Integer> queue = new LinkedList<Integer>();
        queue.push(0);
        while(queue.size()>0){
            int now = queue.pop();
            visited[now]=true;
            //ArrayList<Integer> list = table.get(now);
            ArrayList<Integer> list = table[now];

            for(int i=0; i<list.size(); i++){
                int idx = list.get(i);
                if(connections[idx][0]==now){
                    if(!visited[connections[idx][1]]){
                        ans++;
                        queue.push(connections[idx][1]);
                    }
                }else{
                    if(!visited[connections[idx][0]]){
                        queue.push(connections[idx][0]);
                    }
                }
            }
        }
        return ans;
    }
}

File: 1466/reorder-routes-to-make-all-paths-lead-to-the-city-zero.py
# LeetCode 1466. Reorder Routes to Make All Paths Lead to the City Zero
# n 個城市，有n-1條路相接，原本 connections 裡「城市a」單向到「城市b」
# 但你需要將一些路「轉向」，讓「每條路都會到城市0」即「條條大路通羅馬」的意思
class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        # 技巧：可反過來想，從「城市0」出發，要能到全部的城市。再相反即可
        path = defaultdict(list)  # 「正常」的路
        path2 = defaultdict(list)  # 「反過來」的路
        for a,b in connections:
            path[a].append(b)  # 正常的路
            path2[b].append(a)  # 反過來的路
        self.ans = 0
        visited = set()
        def helper(i):  # 從城市 i 出發
            visited.add(i)  # 「標示」走過i不再走
            for b in path[i]:  # 可「順利」走到的城市
                if b not in visited:  # 還沒走過的，就走
                    self.ans += 1  # 反過來思考，才能到
                    helper(b)  # 函式呼叫函式
            for b in path2[i]:  # 「反過來」的城市
                if b not in visited:
                    helper(b)
        helper(0)  # 從「城市0」出發
        return self.ans  # 反過來思考
